cmd/sink/expr/cel.go

```diff
@@ -0,0 +1,211 @@
+func EmptyFilters() *Filters {
+	return &Filters{
+		archive: make(map[string]cel.Program),
+		delete:  make(map[string]cel.Program),
+	}
+}
+
+// Creates a Filters struct from the path to a directory where a ConfigMap was mounted. If path is empty string or path
+// does not exist, it returns a Filters struct with empty archive and delete slices. It will attempt to create all the
+// cel programs that it can from the ConfigMap. Any errors that are encountered are wrapped together and returned. Even
+// if this function returns an error, the Filters struct returned can still be used and will not be nil.
+func NewFilters(path string) (*Filters, error) {
+	var errList []error
+	filters := &Filters{
+		archive: make(map[string]cel.Program),
+		delete:  make(map[string]cel.Program),
+	}
+	exists, err := files.PathExists(path)
+	if err != nil {
+		return filters, fmt.Errorf("cannot determine if ConfigMap is mounted at path %s: %s", path, err)
+	}
+	if path == "" || !exists {
+		return filters, fmt.Errorf("cannot create Filters. ConfigMap is not mounted at path %s", path)
+	}
+	filterFiles, err := files.DirectoryFiles(path)
+	if err != nil {
+		return filters, fmt.Errorf(
+			"cannot create Filters. Could not read files created from ConfigMap mounted at path %s: %s",
+			path,
+			err,
+		)
+	}
+	for namespace, filePath := range filterFiles {
+		resourcesBytes, err := os.ReadFile(filePath)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		resources := []v1alpha1.KubeArchiveConfigResource{}
+		err = yaml.Unmarshal(resourcesBytes, &resources)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		for _, resource := range resources {
+			namespaceGvk := NamespaceGroupVersionKind(namespace, resource)
+			archiveExpr, err := CreateCelExpr(resource.ArchiveWhen)
+			if err != nil {
+				errList = append(errList, err)
+			} else {
+				filters.archive[namespaceGvk] = *archiveExpr
+			}
+			deleteExpr, err := CreateCelExpr(resource.DeleteWhen)
+			if err != nil {
+				errList = append(errList, err)
+			} else {
+				filters.delete[namespaceGvk] = *deleteExpr
+			}
+		}
+	}
+
+	return filters, errors.Join(errList...)
+
+}
+
+// Returns whether obj needs to be archived. Obj needs to be archived if any of the cel programs in archive return true
+// or if obj needs to be deleted. If obj is nil, it returns false
+func (f *Filters) MustArchive(ctx context.Context, obj *unstructured.Unstructured) bool {
+	if obj == nil {
+		return false
+	}
+	mustArchive := f.MustDelete(ctx, obj)
+	if mustArchive {
+		return mustArchive
+	}
+	globalGvk := GlobalGroupVersionKind(obj)
+	namespaceGvk := NamespaceGroupVersionKindFromObject(obj)
+	program, exists := f.archive[globalGvk]
+	if exists {
+		mustArchive = executeCel(ctx, program, obj)
+	}
+	if mustArchive {
+		return mustArchive
+	}
+	program, exists = f.archive[namespaceGvk]
+	if exists {
+		mustArchive = executeCel(ctx, program, obj)
+	}
+	return mustArchive
```

Comment:
Why are you evaluating the global archive and the namespace archive expressions separately?  You can just `or` them and do one evaluation.

Reply:
After check if the global CEL expression and the namespaced CEL expression exist I could do

```
return executeCel(ctx, globalProgram obj) || executeCel(ctx, namespacedProgram, obj)
```

If you instead mean, taking the string representations of the CEL expressions and doing something like

```
fmt.Sprintf("%s || %s", namespacedCelExprStr, globalCelExprStr)
```

and then compiling that, is it makes updates more difficult. When you mount a configmap every key under `Data` becomes a file and value for that key is the content of that file. 

```
...
Data:
  _global: status.completedTimestamp != ''
```
becomes a file named `_global` and the content of that file is `status.completedTimestamp != ''`

With having the global expressions separate I can treat every file the same. If it is updated, I read in its content and do `yaml.Unmarshal` to convert it into `[]KubeArchiveConfigResource`. If `or` the global expressions of a specific `Kind` to every namespaced expression of that `Kind` I now need to

1. treat `_global` differently than every other file, which will make the live updating logic even more complicated than it already is
2. find every namespace expression that matches the `Kinds` found in `_global` and recompile **every single one of those CEL expressions** which will pause the execution of all cloudevents the sink receives while this is happening

Reply:
The `_global` namespace should look exactly like every other namespace in the `ConfigMap`.  When it's read, its contents should be passed to `yaml.Unmarshal` and converted to a `[]KubebeArchiveConfigResource`.

Here are the steps you can go through:

1. Convert the global structure to a map of `Kind` to expression.  Or, if possible, convert it and add it to the filter, as the filter is already a map of `Kind` to CEL program.
1. Process all the other namespaces.  For each `Kind` value in the `[]KubebeArchiveConfigResource`, see if there is a match in `_global`. If so, `or` it in right then.  What you or it with depends on what works. If you need the string, then create the map described in the previous step.  If you can `or` another CEL program, then use the converted value already in the filter.

> With having the global expressions separate I can treat every file the same. 

This makes no sense to me.  I think you are already treating `_global` differently (and it's one of those files) if you aren't using `yaml.Unmarshall` on it and converting it to a filter.

Reply:
This was more straight forward to code so I did it this way to see if this all would work. I'm working on formatting the expression like this

if the `_global` contains a`KubeArchiveConfigResource` has a kind that matches the kind of a namespaced KubeArchiveConfigResource they will be `or`ed in the following way (making sure that `deleteWhen`s only get `or`ed with `deleteWhen`s. The same with `archiveWhen`s)

```bash
( <global cel expression> ) || ( <namespaced cel expression> )

Reply:
This logic has been implemented. PTAL